### ISSUE TITLE
Exchange heartbeat subscriptions (streaming)

### ISSUE THREAD

**Issue by earce** [knowm/XChange#3523]
Having an issue with CoinbasePro where disconnects on the web socket were getting triggered every so often after some period of inactivity (this due to the fact that on my user channel I don't produce too many trades so messages can be sparse).

I solved this by effectively copying out `CoinbaseProStreamingService.java` amongst other classes to subscribe to CoinbasePro's heartbeat channel which receives heartbeats once a second and prevents web socket from dying. 

I am now seeing a similar issue with Bitstamp (albeit not as frequently) on their regular trades channel with some symbols. Bitstamp does not have a heartbeat channel but suggests using web sockets ping pong mechanism. Sending a way to send a non text based web socket frame is not exposed in the `NettyStreamingService.java` class. 

In short, I would like to make two changes (or more) to the library:

1). Allow users the ability to subscribe to heartbeat channel as an exchange spec config for CoinbasePro (or have it by default subscribe by don't pass heartbeat messages back to user of class-- Kraken does this).

2). Either expose method in `NettyStreamingService.java` to write non text based messages or just a ping method so that users can trigger web socket ping with some interval.

**Comment by badgerwithagun** [comment]
`NettyStreamingService` already responds to ping with pong. I assume Bitstamp are suggesting that client should do the ping? If yes, then I think an overload of `sendMessage` that takes a generic Netty frame seems like a reasonable addition.

I can't think of a good reason to make the Coinbase Pro behaviour optional. It has very low cost and adds reliability for everyone. Once a second seems a little too frequent though. How often does it need to be to prevent disconnect? Half that seems safe enough.

For both Bitstamp and Coinbase, there's an existing pattern of just using an `Observable.interval` subscribed on `connect` and disposed on `disconnect`. See https://github.com/knowm/XChange/blob/develop/xchange-stream-bitfinex/src/main/java/info/bitrich/xchangestream/bitfinex/BitfinexStreamingService.java#L99.

**Comment by earce** [comment]
The only documentation Bitstamp has is here https://www.bitstamp.net/websocket/v2/ nothing about this however, is unique to Bitstamp this is just part of the web socket spec. And yes I am just trying to expose a way to send a ping periodically to Bitstamp to keep the connection alive.

The only reason I would think this being optional would be someone wanting to operate in a low-garbage environment where the heartbeat messages may create unwanted garbage (seems unlikely as I don't see low/no GC being a bit part of the library). As far as heartbeat rate-- that isn't controlled client side, you simply open a subscription with CoinbasePro on its heartbeat channel and messages are sent from Coinbase. I had observed heartbeat messages coming about once a second when testing https://docs.pro.coinbase.com/#the-heartbeat-channel.

The other bit, from testing it looked like the disconnect came from the client (xchange) not from the exchange. I suspect this because `public void channelInactive(ChannelHandlerContext ctx)` was triggered after a period of time without messages from the CoinbasePro. 
 
I am not following how the observable patten you described ties into this?

**Comment by mdvx** [comment]
try this: add to BitstampStreamingService, I thought I had done this already for Bitstamp.

This causes a Ping/Pong Frame to be send (not quite the same as message), and I think this should be the default implementaion for NettyStreamingService anyway.

`  @Override
  protected void handleIdle(ChannelHandlerContext ctx) {
    ctx.writeAndFlush(new PingWebSocketFrame());
 }`

No guarantee, totally untested tonight

**Comment by earce** [comment]
I think this may actually be the best way to fix this (if testing goes well) and a good default implementation.

Also sounds much cheaper then opening a subscription to the heartbeat channel for CoinbasePro.

**Comment by earce** [comment]
Hi everyone,

Implemented the change suggested by @mdvx and works as expected https://github.com/knowm/XChange/pull/3526.

Did have to default idle time so that `protected void handleIdle(ChannelHandlerContext ctx)` would be triggered.

**Comment by badgerwithagun** [comment]
Nice to see the idle feature I added a while ago get used for something! Seems sensible.

**Comment by badgerwithagun** [comment]
Only thing is that now client exchanges need to call super to avoid losing the ping/pong behaviour. Have you checked that there aren't any doing so?

**Comment by earce** [comment]
There are only two exchanges that override it Lgo + Gemini, but the override does the same thing as the feature we just added.
